### ISSUE TITLE
No IME support？

### ISSUE THREAD

**Issue by jinyuanxie** [google/xi-mac#18]
I use XiEditor along with chinese input method, but the window of input method didn't follow the cursor.
![2017-04-18_12-38-10](https://cloud.githubusercontent.com/assets/15538025/25114942/1fb249ca-2435-11e7-81a5-ab1bd951b13a.gif)

**Comment by pocket7878** [comment]
https://github.com/google/xi-editor/pull/70

**Comment by pocket7878** [comment]
Hmm...
Maybe some changes affects to `func firstRect(forCharacterRange aRange: NSRange, actualRange: NSRangePointer?) -> NSRect`'s implementation

**Comment by jinyuanxie** [comment]
@pocket7878 You are right, it always report like that `XiEditor[8625:49370] (0.0, 0.0, 0.0, 0.0)`, i'm trying to fix this issue.

**Comment by pocket7878** [comment]
@jinyuanxie Thank you!

**Comment by pocket7878** [comment]
It seems `cursorPos` didn't updated in draw method

**Comment by jinyuanxie** [comment]
@pocket7878  Thank you! I will create a PR later, but there are another problem.
How should we handle the situation that have multi-cursor ?

**Comment by pocket7878** [comment]
Maybe we should define things like a "active cursor".

**Comment by jinyuanxie** [comment]
Maybe we should wait backend for supporting "Multiple Cursors" https://github.com/google/xi-editor/issues/188

**Comment by cmyr** [comment]
this will definitely be revisited when multi-cursor lands, but I'm not sure how IME would even work if you're inserting text in different contexts simultaneously. If you have any ideas I'm very curious! :)

**Comment by tsekityam** [comment]
I would like to further improve this behaviour. 

The input box should not always follows the cursor. The box should appears at the position of active cursor, then the box's position should be fixed until the input is completed.

That's how the box works in Xcode (and most other editor, if not all), 
![xcode](https://cloud.githubusercontent.com/assets/3814422/25173043/6082928e-2525-11e7-8409-2bb24642ce99.gif)

However, in Xi, the position of the box follows the cursor from time to time,
![xi](https://cloud.githubusercontent.com/assets/3814422/25173046/63bf37a4-2525-11e7-97a1-4dd3e31e8320.gif)

Current behaviour is a bit annoying because It makes me harder to search for a word in the input box.

I will make a PR later tonight.

**Comment by jinyuanxie** [comment]
@tsekityam This behavior may fit XCode by choosing value shown below to update `cursorPos` orderly
- `markRangeStart`
- `selectedRangeStart`
- `cix`

Or maybe we should define their priority?

**Comment by cmyr** [comment]
@tsekityam @pocket7878 @jinyuanxie  I believe you're right, and the implementation of `NSInputClient` is not currently correct. Happy to try and help with a fix. If you have any questions, some of us are in channel #xi on irc.mozilla.org.

**Comment by tsekityam** [comment]
I will try fixing the implementation of the protocol in this weekend.

**Comment by tsekityam** [comment]
I am almost done, and here is [the branch of my patch](https://github.com/tsekityam/xi-mac/tree/patch_review_nstextinputclient_protocol_implementation)

But there are two problems because of the "insert" requests are async, and I want to make it sync.

Some IME, such as Cangjie, will offer predictive completion.

After the marked text, **日日日**,  is replaced by the word **晶**; the input box will still be there and show the predictive suggestion of the word **晶**.

There's how it looks like in EditText:
![textedit predictive completion demo](https://cloud.githubusercontent.com/assets/3814422/25304772/7ae99f50-27a0-11e7-855c-33455a7d2587.gif)


In TextEdit, the position of predictive completion input box is calculated from the new selection range. The new selected range should be `{pos of 晶 + 1, 0}`. However, I found that the word **晶** is not inserted immediately after `insert` request is sent because of the request is async. As a result, the selected range used to calculate the position of the box will never be `{pos of 晶 + 1, 0}`.

Here is how the completion looks like in my branch
![xi predictive completion demo gif](https://cloud.githubusercontent.com/assets/3814422/25304777/958b1348-27a0-11e7-9e83-c2cd52e35ba2.gif)
As you can see the completion box is always appeared next to the previous selected range, but not the latest selected range.

If the insertion is not async, I have no way to calculate the position of completion box with the selected range returned from backend

Besides this issue, the patch is ready for review.

**Comment by tsekityam** [comment]
I just notice that the predictive completion offered in Xi is different from EditText, even if the input are the same **晶**, I will take a look at this.

**Comment by jinyuanxie** [comment]
`selectedRange()` and `markedRange()` seems report the right value.

In my opinion, the previous implement, the `_markedRange` is **relative** position of the `cursorPos`. But the `_selectedRange` is wrong. This implement, the `_markedRange` and `_selectedRange` is **relative** position of the **first line of the viewport**, not the whole document. 

Although the `NSTextInputClient` document point put that should be **absolute** position of the document, but I think use **relative** position will reduce some calculations.

**Comment by tsekityam** [comment]
The value of the ranges will be used in other function. e.g. they will be set as the parameter of `func firstRect(forCharacterRange range: NSRange, 
   actualRange: NSRangePointer?) -> NSRect`, when the app wants to show the IME input box.

If the value of the range are relative position, then `firstRect` function will receive a relative range, which is not expected. As a result, we have to ignore the parameter being set to the `firstRect` when we implement it.

That's ok if the IME input box is the one and only one caller of `firstRect`, `selectedRange` and `markedRange`, however, it may not be true. 

For example, we may call `firstRect` when we want to draw a autocompletion box or a popover on a range that is not selected or marked.
 
Here is one of such popover(!?), that is not drawn on non selected/marked area, in Xcode.
<img width="316" alt="screen shot 2017-04-22 at 11 15 38 pm" src="https://cloud.githubusercontent.com/assets/3814422/25305629/b321166c-27b1-11e7-8561-0fb99a7b835e.png">

User may expect that the base we have implemented the protocol correctly when he extends the view. If we don't implement the protocol correctly and he doesn't notice that, then he may be in trouble.

I think it is bad to implement something not following the document by design.

**Comment by tsekityam** [comment]
By the way, the cause of the strange predictive completion suggestion is that, we didn't implement  protocol correctly.

The IME is using the result of protocol functions, `- (NSAttributedString *)attributedSubstringForProposedRange:(NSRange)range actualRange:(NSRangePointer)actualRange;` or `- (NSAttributedString *)attributedString;` to get the word for prediction.

If any of these functions is implemented, then the predictive completion should work.

**Comment by jinyuanxie** [comment]
Here is my thought about IME support:

* Consider the __relative__ or __absolute__ position should be used.
* How to handle multi-selection/cursor?
* Should we use temporary variable to storage the marked text instead of call backend to do insertions. The related behavior can be described as follow:
    1. I selected some text, then type something using InputMethod.
    2. I want to discard the typed text, and go back previous state.
    3. Because currently we call backend to do insert in `setMarkedText`, we should apply undo more than once to back to previous state.

Also, we should reduce the calculations in `draw` function which will be call frequently.

**Comment by tsekityam** [comment]
* How to handle multi-selection/cursor?
    Mark last added selection/cursor as main selection/cursor, and return it from `selectedRange()`. 

    If there is marked text to be inserted, then we drop all non-main selection/cursor and only set marked text to main selection/cursor.

    **TL;DR:** don't support multi selection/cursor in IME input

    ***
    P.S. Atom supports IME input with multiple cursor, however, I don't like their implementation. The IME box position varies while typing, it's hard to locate a word in the box
![ime with multi cursor in atom](https://cloud.githubusercontent.com/assets/3814422/25311845/b25d7fb8-283d-11e7-88e8-54e42d81fd42.gif)

**Comment by jansol** [comment]
For multiple cursors, the easiest and most intuitive behaviour would probably be showing the IME box at the cursor that was placed last.

**Comment by jansol** [comment]
Any progress on this? Might be worth opening at least a WIP pull request so others can test and comment on it better.

As for positioning the suggestions, at least for Japanese the segment being edited is further split into subsegments and the suggestion box is aligned with the start of the current subsegment (source: TextEdit, Xcode and Pages). For the multi-cursor scenario I'd stick with my previous suggestion of following the cursor that was placed last, i.e. align with the beginning of the current subsegment starting from the last cursor's position.

**Comment by cmyr** [comment]
We had this working before the openGL stuff. I believe it isn't working again now, or at least that the IME window is being presented incorrectly.

**Comment by jansol** [comment]
Currently the IME window is stuck to the bottom left corner so yeah, not presented correctly.
